ENH: Add na_value argument to DataFrame.to_numpy

dsaxton opened the pull request:
- [x] closes #33820
- [x] tests added / passed
- [x] passes `black pandas`
- [x] passes `git diff upstream/master -u -- "*.py" | flake8 --diff`
- [x] whatsnew entry

dsaxton committed:
ENH: Add na_value argument to DataFrame.to_numpy

simonjayhawkins commented on pandas/core/frame.py:
```diff
@@ -1337,6 +1344,10 @@ def to_numpy(self, dtype=None, copy: bool = False) -> np.ndarray:
         result = np.array(self.values, dtype=dtype, copy=copy)
+
+        if na_value is not lib.no_default:
+            result[isna(result)] = na_value
```
```
>>> df = pd.DataFrame({"A": [1, 2, None]})
>>> df.to_numpy(na_value=pd.NA)
Traceback (most recent call last):
  File "<stdin>", line 1, in <module>
  File "C:\Users\simon\pandas\pandas\core\frame.py", line 1349, in to_numpy
    result[isna(result)] = na_value
TypeError: float() argument must be a string or a number, not 'NAType'
>>>
```
do we want a more user friendly message, or perhaps validate the fill_value.

jorisvandenbossche replied:
That will be difficult in general, I think (since it depends on the behaviour of numpy on what it accepts as value to set to the array). 
For this specific case, `pd.NA` is basically only possible for object dtype (so that could be checked specifically. Although it is also not the typical use case for `to_numpy`, since you typically want to get rid of pd.NA, since that doesn't work nicely with numpy).

Now, for an extension array, you already get a slightly better error message:

```
In [18]: pd.array([1, 2, None]).to_numpy(float, na_value=pd.NA)   
---------------------------------------------------------------------------
ValueError                                Traceback (most recent call last)
<ipython-input-18-3c4196341619> in <module>
----> 1 pd.array([1, 2, None]).to_numpy(float, na_value=pd.NA)

~/scipy/pandas/pandas/core/arrays/masked.py in to_numpy(self, dtype, copy, na_value)
    145             ):
    146                 raise ValueError(
--> 147                     f"cannot convert to '{dtype}'-dtype NumPy array "
    148                     "with missing values. Specify an appropriate 'na_value' "
    149                     "for this dtype."

ValueError: cannot convert to '<class 'float'>'-dtype NumPy array with missing values. Specify an appropriate 'na_value' for this dtype.
```

simonjayhawkins commented on pandas/core/frame.py:
```diff
@@ -1305,6 +1307,11 @@ def to_numpy(self, dtype=None, copy: bool = False) -> np.ndarray:
             a copy is made, even if not strictly necessary.
+        na_value : Any, optional
+            The value to use for missing values. The default value depends
+            on `dtype` and the type of the array.
```
what happens with mixed dataframes?

jorisvandenbossche replied:
Then the same value is used for all dtypes.

(in principle you could imagine this keyword being a dict to specify a value per column, but let's leave that for later if there is actually a use case / demand for this)

simonjayhawkins reviewed (commented):
Thanks @dsaxton maybe some tests for error messages and mixed dataframes

jorisvandenbossche commented on pandas/core/frame.py:
```diff
@@ -1337,6 +1344,10 @@ def to_numpy(self, dtype=None, copy: bool = False) -> np.ndarray:
                [2, 4.5, Timestamp('2000-01-02 00:00:00')]], dtype=object)
         """
         result = np.array(self.values, dtype=dtype, copy=copy)
```
For some dtypes, it will not work to first convert to an array of that dtype, and only afterwards to replace missing values. 
For example, with Int column with missing value, the initial step to convert to a float array already fails:

```
In [22]: pd.DataFrame({'a': [1, 2, None]}, dtype="Int64").to_numpy(float) 
---------------------------------------------------------------------------
TypeError                                 Traceback (most recent call last)
<ipython-input-22-2847bf76b2be> in <module>
----> 1 pd.DataFrame({'a': [1, 2, None]}, dtype="Int64").to_numpy(float)

~/scipy/pandas/pandas/core/frame.py in to_numpy(self, dtype, copy)
   1335                [2, 4.5, Timestamp('2000-01-02 00:00:00')]], dtype=object)
   1336         """
-> 1337         result = np.array(self.values, dtype=dtype, copy=copy)
   1338         return result
   1339 

TypeError: float() argument must be a string or a number, not 'NAType'
```

So I think that we will need to rework the "DataFrame -> array" conversion a bit (so fix this deeper in the internals). We need to ensure that for EAs, we call `to_numpy` with the kwargs passed through, before combining the different columns in a single array.

dsaxton replied:
@jorisvandenbossche Updated to do this in a slightly better way. The below test will fail however as this approach can't guarantee non-copying behavior:
https://github.com/pandas-dev/pandas/blob/085752f95adb625a43aef64481be56a393e8732a/pandas/tests/frame/test_api.py#L363

jorisvandenbossche commented on pandas/tests/base/test_conversion.py:
```diff
@@ -410,3 +410,13 @@ def test_to_numpy_kwargs_raises():
+    data = pd.DataFrame({"a": [1, 2, 3], "b": [1, 2, None]})
+    result = data.to_numpy(dtype=dtype, na_value=na_value)
+    expected = np.array([[1, 1], [2, 2], [3, na_value]], dtype=dtype)
+    tm.assert_numpy_array_equal(result, expected)
```
Can you add some more tests for a few "realistic" cases? Like a dataframe with nullable integer array (as mentioned above I think this will still fail) converting to float with np.nan as missing value, or converting to int with -1 as missing value.

dsaxton committed:
Add some tests

dsaxton committed:
Issue num

dsaxton committed:
A little better

dsaxton committed:
Merge remote-tracking branch 'upstream/master' into dataframe-to-numpy

jreback commented on pandas/core/frame.py:
```diff
@@ -1336,7 +1343,12 @@ def to_numpy(self, dtype=None, copy: bool = False) -> np.ndarray:
+            self[c].to_numpy(dtype=dtype, na_value=na_value, **kwargs)
+            for c in self.columns
+        ]
+        result = np.array(arr).T
```
pass thru copy=, though I think this may actually *always* copy

jreback replied:
we should add a test for an all float frame to see if it actually does

jreback replied:
we *could* a special case for a single dtype (maybe)

dsaxton replied:
That's tested here and indeed copy=False doesn't do anything after this change:
https://github.com/pandas-dev/pandas/blob/085752f95adb625a43aef64481be56a393e8732a/pandas/tests/frame/test_api.py#L363
I think this is okay since now the copy behavior isn't guaranteed anyways when copy=False (would want to explicitly state this in the docstring probably)?

jorisvandenbossche replied:
We should preserve the no-copy behaviour for a single block, I would say (so the existing test keeps working). 

@dsaxton we should maybe only do the column-wise approach as you do above when there is at least one extension dtype column (then we are sure the dataframe is not a single 2D block)

jorisvandenbossche commented on pandas/core/frame.py:
```diff
@@ -1336,7 +1343,12 @@ def to_numpy(self, dtype=None, copy: bool = False) -> np.ndarray:
+        arr = [
+            self[c].to_numpy(dtype=dtype, na_value=na_value, **kwargs)
+            for c in self.columns
+        ]
```
you can use `self._iter_column_arrays()` here, to avoid creating intermediate Series objects

jorisvandenbossche replied:
Although, that also returns numpy arrays, which makes it more complicated to support na_value for those ..

dsaxton commented:
Another interesting issue is the behavior for mixed numeric and non-numeric data types. In the docstring we say that these will promote to object, which does happen but apparently numpy decides to convert the datetime to int beforehand anyways (we get the expected output if we explicitly ask for object):
```python
[ins] In [1]: df = pd.DataFrame({"A": [1, 2], "B": [3.0, 4.5]})                                                                                                                                              

[ins] In [2]: df['C'] = pd.date_range('2000', periods=2)                                                                                                                                                     

[ins] In [3]: df.to_numpy()                                                                                                                                                                                  
Out[3]: 
array([[1, 3.0, 946684800000000000],
       [2, 4.5, 946771200000000000]], dtype=object)

[ins] In [4]: df.to_numpy(dtype=object)                                                                                                                                                                      
Out[4]: 
array([[1, 3.0, Timestamp('2000-01-01 00:00:00')],
       [2, 4.5, Timestamp('2000-01-02 00:00:00')]], dtype=object)
```

jorisvandenbossche commented:
So I think you need to deal with this deeper into the internals, as I mentioned before. It is there that the custom pandas logic to decide on the resulting dtype is done (which is indeed different from numpy's result dtype)

jorisvandenbossche commented:
Check BlockManager.as_array -> this also already handles the case of a single Block (to keep the no copy behaviour there).

I think we can expand that method to support na_value

dsaxton committed:
as_array

dsaxton committed:
Merge remote-tracking branch 'upstream/master' into dataframe-to-numpy

dsaxton committed:
Black

dsaxton committed:
More black

jorisvandenbossche commented on pandas/core/internals/managers.py:
```diff
@@ -781,14 +781,16 @@ def copy_func(ax):
         return res
 
-    def as_array(self, transpose: bool = False) -> np.ndarray:
+    def as_array(self, transpose: bool = False, na_value=lib.no_default) -> np.ndarray:
```
I think we should also pass `dtype` here, to avoid eg fist converting ints/floats to object and then converting back to numeric dtype in `DataFrame.to_numpy`

jorisvandenbossche commented on pandas/core/internals/managers.py:
```diff
@@ -808,6 +810,9 @@ def as_array(self, transpose: bool = False) -> np.ndarray:
             arr = self._interleave()
 
+        if na_value is not lib.no_default:
+            arr[isna(arr)] = na_value
```
Suggestion: pass through this `na_value` to `_interleave`, and then when looping over the blocks, you can check for ExtensionBlocks and then call `to_numpy` with the appropriate dtype and na_value. 

The lines above will need to be kept for non-EAs, but I think for EAs, we want to use their `to_numpy` implemenation, as this can potentially be more efficient in depending on the dtype/na_value combination

dsaxton replied:
Pushed a change to make this update. In the single extension block case we're still doing a conversion to object (got several odd failures when I tried making the associated edit there) but I'm thinking this is already going to be a pretty inexpensive case anyways

jorisvandenbossche replied:
> In the single extension block case we're still doing a conversion to object 

I think in the above `is self._is_single_block` checks, you can add an additional case for single ExtensionBlock (like there is already a check for single datetimetz block)

dsaxton replied:
Got mixed up by the "blocks are transposed" thing and was missing a reshape, should be working now 🤞

dsaxton committed:
to_numpy for ExtensionBlock

dsaxton committed:
Merge remote-tracking branch 'upstream/master' into dataframe-to-numpy

dsaxton committed:
dtype hack

dsaxton committed:
Merge remote-tracking branch 'upstream/master' into dataframe-to-numpy

dsaxton commented on pandas/core/frame.py:
```diff
@@ -1336,7 +1343,15 @@ def to_numpy(self, dtype=None, copy: bool = False) -> np.ndarray:
+            transpose=self._AXIS_REVERSED, dtype=dtype, na_value=na_value
+        )
+
+        if copy or result.dtype is not dtype:
+            # FIXME: The dtype conversion is potentially needed since the
+            #  single extension block case is not correctly cast above
+            result = np.array(result, dtype=dtype, copy=copy)
```
Not happy with this dtype check, but otherwise there's a regression in the single extension block case. Open to ideas / help on why we can't just do this via to_numpy around where we call get_values in as_array (several tests in frame/methods/test_append.py fail for instance)

jreback replied:
yeah this should be handled in .as_array itself

dsaxton committed:
reshape

dsaxton committed:
Merge remote-tracking branch 'upstream/master' into dataframe-to-numpy

dsaxton committed:
Check for NA first

Otherwise this can raise for an all integer DataFrame without
any missing values

dsaxton committed:
Cast non-extension single block

dsaxton committed:
Test nit

jorisvandenbossche commented on pandas/core/frame.py:
```diff
@@ -1305,6 +1307,11 @@ def to_numpy(self, dtype=None, copy: bool = False) -> np.ndarray:
             a copy is made, even if not strictly necessary.
+        na_value : Any, optional
+            The value to use for missing values. The default value depends
+            on `dtype` and the type of the array.
```
```suggestion
            on `dtype` and the dtypes of the DataFrame columns.
```

? (not sure what was meant here originally)

jorisvandenbossche commented on pandas/core/frame.py:
```diff
@@ -1336,7 +1343,13 @@ def to_numpy(self, dtype=None, copy: bool = False) -> np.ndarray:
+        )
+
+        if copy:
+            result = result.copy()
```
It would be nice to only copy if `result` is actually not yet a copy. (but that might maybe easier to handle inside `as_array` as well?) Now this is not the most pressing issue

jorisvandenbossche commented on pandas/core/internals/managers.py:
```diff
@@ -803,19 +809,34 @@ def as_array(self, transpose: bool = False) -> np.ndarray:
+        if na_value is not lib.no_default:
+            na_mask = isna(arr)
+            if na_mask.any():
+                arr[na_mask] = na_value
```
Actually, if ``arr`` is not a copy at this point, this will modify the original dataframe (eg I assume this will happen with a uniform float dataframe and you do eg `df.to_numpy(na_value=0)`). Can you check if that indeed is the case and add a test for this to ensure the original dataframe didn't change in the single block case?

dsaxton replied:
That's a good point, it looks like the original DataFrame is _not_ mutated (at least in the case you mention)

dsaxton replied:
Now that we may be copying multiple times, I think we should probably move the copy into this function and make sure that doesn't happen?

jorisvandenbossche replied:
Yes, that's probably a good idea

dsaxton committed:
Update pandas/core/frame.py

Co-authored-by: Joris Van den Bossche <jorisvandenbossche@gmail.com>

dsaxton committed:
side effect test

dsaxton committed:
Merge remote-tracking branch 'upstream/master' into dataframe-to-numpy

jorisvandenbossche commented on pandas/tests/base/test_conversion.py:
```diff
@@ -444,3 +444,11 @@ def test_to_numpy_dataframe_single_block(data, expected):
+    # https://github.com/pandas-dev/pandas/issues/33820
+    result = pd.DataFrame(np.array([1.0, 2.0, np.nan]))
+    expected = pd.DataFrame(np.array([1.0, 2.0, np.nan]))
+    result.to_numpy(dtype=float, na_value=0.0)
```
```suggestion
    result.to_numpy(na_value=0.0)
```

I think this might trigger editing the original

dsaxton replied:
Ah, you're right. I guess we do need to check for this explicitly.

dsaxton committed:
Update pandas/tests/base/test_conversion.py

Co-authored-by: Joris Van den Bossche <jorisvandenbossche@gmail.com>

dsaxton committed:
Copy

dsaxton committed:
Copy less

dsaxton committed:
Merge remote-tracking branch 'upstream/master' into dataframe-to-numpy

dsaxton commented on pandas/tests/frame/test_api.py:
```diff
@@ -365,7 +365,7 @@ def test_to_numpy_copy(self):
         assert df.values.base is arr
         assert df.to_numpy(copy=False).base is arr
-        assert df.to_numpy(copy=True).base is None
+        assert df.to_numpy(copy=True).base is not arr
```
This had to change because now the output array is a view, but only on the transpose of itself (which is a copy of the original data)

jorisvandenbossche commented on pandas/core/internals/managers.py:
```diff
@@ -823,10 +831,11 @@ def as_array(
             arr = self._interleave(dtype=dtype, na_value=na_value)
 
+        if copy or na_value is not lib.no_default:
+            arr = arr.copy()
```
When `interleave` is called, that will always be already a copy, so it would be nice to avoid a second copy there as well

Maybe the above if/else can be reorganized slightly, like:

```
if self._is_single_block or not self.is_mixed_type:
  if self.blocks[0].is_datetimetz:
    ...
  elif self.blocks[0].is_extension:
    ...
  else:
    ....
  if copy or na_value is not lib.no_default:
    arr = arr.copy()
else:
  arr = self._interleave(dtype=dtype, na_value=na_value)
```

dsaxton replied:
Good idea. Not sure if it's "bad style" but is another option simply to create a "should_copy" variable and update it for the interleave case? Might be a bit more explicit what we're thinking and helps avoid nested conditionals.

jorisvandenbossche replied:
Such a variable is also fine, whatever is easiest to implement / you prefer

dsaxton committed:
should_copy

dsaxton committed:
Merge remote-tracking branch 'upstream/master' into dataframe-to-numpy

jorisvandenbossche commented on pandas/core/internals/managers.py:
```diff
@@ -798,24 +812,43 @@ def as_array(self, transpose: bool = False) -> np.ndarray:
         elif self._is_single_block or not self.is_mixed_type:
             arr = np.asarray(self.blocks[0].get_values())
+            if dtype:
+                arr = arr.astype(dtype)
```
```suggestion
                arr = arr.astype(dtype, copy=False)
```

dsaxton committed:
Update pandas/core/internals/managers.py

Co-authored-by: Joris Van den Bossche <jorisvandenbossche@gmail.com>

jreback commented on pandas/core/internals/managers.py:
```diff
@@ -798,24 +812,43 @@ def as_array(self, transpose: bool = False) -> np.ndarray:
             arr = np.empty(self.shape, dtype=float)
             return arr.transpose() if transpose else arr
 
+        should_copy = copy or na_value is not lib.no_default
```
why is this needed?

jorisvandenbossche replied:
It's to avoid making a second copy in certain cases where we can detect this (eg with multiple dtypes (going through `_interleave`), the constructed array is never the original values, so an additional copy is not needed

jreback replied:
then i would just make this
```
copy = copy or na_value is not lib.no_default
```

and add an explanation line

jreback commented on pandas/core/internals/managers.py:
```diff
@@ -798,24 +812,43 @@ def as_array(self, transpose: bool = False) -> np.ndarray:
             # always be object dtype. Some callers seem to want the
             # DatetimeArray (previously DTI)
             arr = self.blocks[0].get_values(dtype=object)
+        elif self._is_single_block and self.blocks[0].is_extension:
```
you might be able to remove the block aboev (is_datetimetz) as that is an extension block already

jorisvandenbossche replied:
Indeed, the `DatetimeArray.to_numpy` behaviour seems correct for the default (having object dtype with timestamps instead of datetime64)

dsaxton replied:
Nice 👍

jreback commented on pandas/core/internals/managers.py:
```diff
@@ -798,24 +812,43 @@ def as_array(self, transpose: bool = False) -> np.ndarray:
         elif self._is_single_block or not self.is_mixed_type:
             arr = np.asarray(self.blocks[0].get_values())
+            if dtype:
+                arr = arr.astype(dtype, copy=False)
```
is this tested / hit at all? (the dtype part)

jorisvandenbossche replied:
The added `test_to_numpy_dataframe_single_block` should cover this line I think

dsaxton committed:
Rename and comment

dsaxton committed:
Merge remote-tracking branch 'upstream/master' into dataframe-to-numpy

dsaxton committed:
Don't special case datetimetz

dsaxton committed:
Merge remote-tracking branch 'upstream/master' into dataframe-to-numpy

jreback commented:
thanks @dsaxton
